Split types.jl in different files to improve readability and ease contribution

pabvald opened the pull request:
## Objective 
Make the code with  the different types and related functions more comprehensible, making it easier for future contributors to extend the existing types or add new types.

## Changes
The `src/types.jl` file (around 1200 lines) is split in different files, where each file contains different types: 
- `chunk.jl`: contains the interface and subtypes of `AbstractCandidateChunks`, i.e. `CandidateChunks` and `MultiCandidateChunks`, and related functions.
- `document_term_matrix.jl`: contains the interface and subtypes of `AbstractDocumentTermMatrix`, i.e.`DocumentTermMatrix` and `SubDocumentTermMatrix`, and related functions.
- `index.jl`: contains the interface of `AbstractDocumentIndex`,  the interface and subtypes of `AbstractChunkIndex`, i.e. `ChunkEmbeddingsIndex`, `ChunkKeywordsIndex` and `SubChunkIndex`, as well as the type `MultiIndex`, and related functions
- `rag_result.jl`: contains the type `RAGResult` and related functions.
- `Types.jl`: includes imports and **shared functions** of both index and chunk types.

The `test/types.jl` has been split accordingly.

## Notes 
This refactoring doesn't affect the namespace of the code previously contained in `types.jl`. There's no new `Types` module. There are no breaking changes. Just pure code cleaning and formatting.

pabvald commented:
The coverage of `types.jl` was also 93.64%, equal to the patch coverage: https://app.codecov.io/gh/JuliaGenAI/RAGTools.jl/commit/d07cfd5206ca9e19f946dd0039df9b23300b9d20/blob/src/types.jl?dropdown=coverage. Which makes sense since both the code and the tests are the same, just in different files.

I think the drop on the total coverage is the result of a bad heuristic used by codecov, which seems to still count the removed `types.jl` file

pabvald commented:
Regarding your feedback, @svilupp: 

1. > Could you please make the type files flat? As previously mentioned, there is a cost to these nested structures in terms of bad imports etc, and no one upvoted this refactor, so let's make it as simple&clear as possible.  Just call the files "types_*.jl" and include them from top level.

   Having a `types/` folder doesn't mean you have a `types` module. I have worked with Julia projects that have modules, and yes, nested imports are a pain. I haven't added a submodule to `RAGTools` and there are no breaking changes. 
  
    Having it at the top level doesn't leave room for code that is common to more than one type e.g. `Base.getindex` or other functions that may be added in the future.

2. > Please add a comment above each inclusion as to what's in each file ✅ 


3. > For the DTM types, it might be better to use the full names, because the important part of the phrase is "term". Document matrix might be misleading ✅

pabvald commented:
@svilupp anything against merging the PR?

Sixzero commented:
Yeah, this last thing would be cool to get fixed, keeping the folder `types` but not importing things from that folder is easier for agents to find the type of a thing, because it doesn't need to read the `types/Types.jl` before it know that the type it is looking for is in `types/rag_results.jl` so it is actually one less redirection.
I wonder if the functions that are in thy Types.jl could be... I mean are those really shared functions? :thinking:

Sixzero commented:
Splitting up types.jl I think is a really good thing to make things RAG friendly actually. 

Recently I just realised chunking over 8k token any file what is around 400-600 line of code is actually to me now seems unpractical. Most embedders only support 8k token, and I think usually assigning one vector to 500 LOC is just not that good representation of something, so yeah keeping files under 500 LOC is also IMO now a good practice to keep if we want to stay RAG friendly. :blush:

pabvald commented:
> I would prefer to not have a file called "types/Types.jl". Would it be possible to import the files directly into the top-level module? It could make it easier for LLMs to locate the right type definitions.

> Yeah, this last thing would be cool to get fixed, keeping the folder types but not importing things from that folder is easier for agents to find the type of a thing, because it doesn't need to read the types/Types.jl before it knows that the type it is looking for is in types/rag_results.jl so it is actually one less redirection.

I have got rid of the `src/types/Types.jl` and `test/types/Types.jl` files. 


> types/chunk.jl - it doesn't feel descriptive to me. Would "candidate_chunks" or candidates be too long?

You are right, `canditate_chunks.jl` is better. I have renamed it. 

> What's the logic for having some definitions in types/Types.jl? What was the rule to include them there versus in the individual-type files? 

> I wonder if the functions that are in the Types.jl are really shared functions? I wonder if we could somehow locate them or hmm... I don't know where they should be. 🤔

I have moved both `Base.view` and `Base.getindex` to `types/index.jl`. They are in the end functions of the indexes, even if they also involve the candidate chunks.

svilupp reviewed (changes requested):
Could you please remove the repeated deps imports in all files in types/ 

I can see `using JSON3` at the top of each, but I'm not sure we have to explicitly re-import it every time?

pabvald commented:
@svilupp  sure, no problem, `JSON it's already imported in `RAGTools.jl` so I have deleted it.
